### ISSUE TITLE
Cannot open Serial Port with an arbitrary (i.e 100000) baud rate

### ISSUE THREAD

**Issue by burlutskiy** [Pi4J/pi4j-v1#89]
The following code returns -2

``` java
fileDescriptor = com.pi4j.wiringpi.Serial.serialOpen("/dev/ttyAMA0", 100000);
```

if you set 38400, 57600, 115200, etc it work perfectly fine, but if it is something in between it doesn't work, which is strange, for example python script works on 100000:

``` python
#!/usr/bin/python
import serial
ser = serial.Serial('/dev/ttyAMA0',100000,timeout =10)
```

Environment:
$ java -version
java version "1.8.0-ea"
Java(TM) SE Runtime Environment (build 1.8.0-ea-b103)
Java HotSpot(TM) Client VM (build 25.0-b45, mixed mode)
$ uname -a
Linux raspberrypi 3.10.25+ #622 PREEMPT Fri Jan 3 18:41:00 GMT 2014 armv6l GNU/Linux

**Comment by savageautomate** [comment]
I believe that may be a limitation of the underlying WiringPi library.  We are limited to only the baud rates that is exposes.

**Comment by burlutskiy** [comment]
I had a discussion with Gordon, so yes it supports only standard baud rates. I have been working on adding support for non-standard baud rate into wiringSerial.c

I'm a bit stuck with an issue:

```
  [sshexec] Connecting to 10.0.1.3:22
  [sshexec] cmd : cd /home/pi/pi4j-temp; make clean; make all
rm -f com_pi4j_jni_I2C.o com_pi4j_jni_Serial.o com_pi4j_wiringpi_Gpio.o com_pi4j_wiringpi_Serial.o com_pi4j_wiringpi_Lcd.o com_pi4j_wiringpi_Shift.o com_pi4j_wiringpi_SoftPwm.o com_pi4j_wiringpi_Spi.o com_pi4j_wiringpi_Gertboard.o com_pi4j_wiringpi_Nes.o com_pi4j_wiringpi_GpioPin.o com_pi4j_wiringpi_GpioUtil.o com_pi4j_wiringpi_GpioInterrupt.o libpi4j.so *~ core tags Makefile.bak
[COMPILE] gcc -c -O3 -Wall -I. -I//include -I/usr/local/include -I. -I//include/linux -I/usr/local/include/linux -Winline -pipe -L/usr/local/lib /usr/local/lib/libwiringPi.a  com_pi4j_jni_I2C.c
gcc: error: /usr/local/lib/libwiringPi.a: No such file or directory
make: *** [com_pi4j_jni_I2C.o] Error 1
```

I have only the following files in that folder

```
pi@raspberrypi ~ $ ls /usr/local/lib/libwiringPi
libwiringPiDev.so      libwiringPiDev.so.2.0  libwiringPi.so         libwiringPi.so.2.0
```

Do you know why it fails?

**Comment by savageautomate** [comment]
Not entirely sure.  The build was pulling wiringPi from a GitHub repo, but I found that repo to be a fork from Gordon's main repo and not always in sync.  I have since updated the build to pull from Gordon's repo and to only build for hard-float ABI.  Make sure to clone the 'develop' branch of Pi4J to try the latest.

**Comment by savageautomate** [comment]
I have implemented this feature in an experimental new branch (version 1.1-SNAPSHOT). 
However, support for custom baud rates is very dependent on your serial port driver.  Many don't support custom baud rates.

---

I have been working on a new serial implementation to properly support a byte level interface for working with the data.  I suggest that you try the experimental build 1.1-SNAPSHOT.  The serial API is completely overhauled and should provide far better results when working with bytes. 

GITHUB BRANCH:
https://github.com/Pi4J/pi4j/tree/serial_enhancements

DOWNLOAD:
http://get.pi4j.com/download/pi4j-1.1-SNAPSHOT.zip

Please let me know if you find any problems ... remember it is very new and experimental :-)

**Comment by savageautomate** [comment]
Added in 1.1-SNAPSHOT build.

**Comment by mtstorm** [comment]
I have tried to use this. I am working on an implementation for the BrickPi API. The BrickPi uses a baudrate of 500,000. The serial does not fail on opening but it is not working. It's not reading after sending data. The link to 'serial_enhancements' is not working so I would like to know where can I find that. I do not use the event driven aproach of the serial interface. First sending some data and than waiting for the data to come back, but there is no data coming back in the buffer (chip sends it back). So there is or a bug in the API or in my code. In order to check my code I have tried to read the javadoc and found something that maybe helps. I have added 'setBufferingDataReceived' in the hope I get some data in de received buffer.

serial.open(Serial.DEFAULT_COM_PORT, 500000);
serial.setBufferingDataReceived(true);

But this was not helping at all, my bug remains. So I was wandering a way to valide if it is getting some data. I have made a simple loopback programm and connected the serial rx to tx on the rpi. This programm works but there is a small problem. The old serial.shutdown has been removed and autocloseable is now on the serial, the programm will not exit unless forcing it with System.exit().

```
public static void main(String[] args)   {
    try (Serial serial = SerialFactory.createInstance();) {
        serial.open(Serial.DEFAULT_COM_PORT, 500000);
        serial.setBufferingDataReceived(true);
        String sendMessage = "This is a loopback test";
        serial.write(sendMessage);

        Thread.sleep(1000);
        byte[] bytesReived = serial.read();
        for (byte b : bytesReived) {
            System.out.println((char)b  +"  value "+ b);
        }
        String receivedMessage = new String(bytesReived);
        System.out.println("receivedMessage := "+receivedMessage);
        System.out.println("Transmission succes "+  sendMessage.equals(receivedMessage));

    } catch (IOException | InterruptedException e) {
        e.printStackTrace();
    } finally {
        //serial thread is not closed!!
        System.out.println("Done");
        //force exit 
        System.exit(-1);
    }
}
```

Three questions for my basic understanding:
Where can I find the link to 'serial_enhancements', since it is not working.
When I do serial.write, will it transmit directly?
I can I terminate/ release the serial interface, and could this be implemented using to try-with resource statement?

I have build and tested with the following version "1.1-20160409.004953-44".

I hope you will find the time to respond because I need this to work. I am happy to help. mtstorm

**Comment by savageautomate** [comment]
Hi @mtstorm 

The serial_enhancements was an earlier code branch that included the proposed modifications to the serial API and implementation.  Its gone now because the proposed changes have been adopted (merged) and are now part of the active codebase in the 'develop' (https://github.com/Pi4J/pi4j/tree/develop/pi4j-core/src/main/java/com/pi4j/io/serial) brach and currently included in 1.1-SNAPSHOT builds.

When you perform a serial write, yes it is transmitted (_almost_) immediately.  No further action should be needed on your part to force data in the transmit buffer out. 

The symptoms you are seeing may come down the to the custom baud rate you are trying to use: 500000.   While Pi4J and the underlying linux serial APIs support the custom baud rate, the hardware UART must also support the custom baud rate.  I have tested higher and custom speeds on external USB to SERIAL adapters, but not on the internal (default) serial port on the GPIO header.  

It looks like it may be possible to drive the default serial port at higher speeds, but may require some additional configuration.  I came across these articles that may be helpful:
http://www.thedevilonholiday.co.uk/raspberry-pi-increase-uart-speed/
https://www.raspberrypi.org/forums/viewtopic.php?f=44&t=73673

Again, I have not tried these customizations, but it look like a promising place to start.

Thanks, Robert
